### PR TITLE
Missing values in filter and smoother for KalmanFilter

### PR DESCRIPTION
**Pull request by raphaelsaavedra**
Addressing #43

### CODE REVIEW

File: src/kalman.jl [hunk 1]
```diff
@@ -33,25 +33,36 @@ function kalman_filter(model::StateSpaceModel, H::Matrix{Typ}, Q::Matrix{Typ}; t
...
-            # Avoid numerical errors by ensuring positivity and symmetry
-            P[:, :, t+1] = (P[:, :, t+1] + P[:, :, t+1]')/2 + 1e-8.*Matrix{Float64}(I, m, m)
-
-            # Checking if steady state was attained
-            if check_steady_state(P[:, :, t+1], P[:, :, t], tol)
-                steadystate = true
-                tsteady     = t
+            a[t+1, :]    = T*a[t, :]
+            P[:, :, t+1] = T*P[:, :, t]*T' + R*Q*R'
+            P[:, :, t+1] = ensure_pos_sym(P[:, :, t+1])
```
Comment (guilhermebodin) [inline src/kalman.jl]: `ensure_pos_sym(T*P[:, :, t]*T' + R*Q*R')`

File: src/kalman.jl [hunk 2]
```diff
@@ -33,25 +33,36 @@ function kalman_filter(model::StateSpaceModel, H::Matrix{Typ}, Q::Matrix{Typ}; t
...
+                a[t+1, :]    = T * a[t, :] + K[:, :, t] * v[t, :]
+                P[:, :, t+1] = P[:, :, t]
+            else
+                F[:, :, t]   = Z[:, :, t] * P[:, :, t] * Z[:, :, t]' + H
+                K[:, :, t]   = T * P[:, :, t] * Z[:, :, t]' * inv(F[:, :, t])
+                a[t+1, :]    = T * a[t, :] + K[:, :, t] * v[t, :]
+                P[:, :, t+1] = T * P[:, :, t] * (T - K[:, :, t] * Z[:, :, t])' + R*Q*R'
+
+                # Avoid numerical errors by ensuring positivity and symmetry
+                P[:, :, t+1] = ensure_pos_sym(P[:, :, t+1])
```
Comment (guilhermebodin) [inline src/kalman.jl]: same here

File: src/kalman.jl (hunk 1 above)
Comment (guilhermebodin) [reply src/kalman.jl]: just an opinion, a matter of style

File: src/utils.jl [hunk 3]
```diff
@@ -39,4 +39,16 @@ Return `true` if steady state was attained with respect to tolerance `tol`, `fal
...
     return maximum(abs.((P_t1 - P_t)./P_t1)) < tol ? true : false
+end
+
+"""
+    ensure_pos_sym(M::Matrix{T}; ϵ::T = 1e-8) where T <: AbstractFloat
+
+Ensure that matrix `M` is positive and symmetric to avoid numerical errors when numbers are small
+"""
+function ensure_pos_sym(M::Matrix{T}; ϵ::T = 1e-8) where T <: AbstractFloat
+    m, n = size(M)
```
Comment (guilhermebodin) [inline src/utils.jl]: n = `LinearAlgebra.checksquare(M)`

File: src/kalman.jl (hunk 1 above)
Comment (raphaelsaavedra) [reply src/kalman.jl]: yeah that's much better, changed in 61744d45d90a964030dba61db26cf85e6be7be8b

File: src/kalman.jl (hunk 2 above)
Comment (raphaelsaavedra) [reply src/kalman.jl]: changed in 61744d45d90a964030dba61db26cf85e6be7be8b

File: src/utils.jl (hunk 3 above)
Comment (raphaelsaavedra) [reply src/utils.jl]: Joaquim suggested we drop this check altogether since it will throw a dimension error anyway, I agreed and removed it

**Pull request merged**